File: backend/app/schemas/tournament.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.badge import BadgeRead

AttemptStatus = Literal["started", "valid", "dnf"]


class TournamentAttemptSubmit(BaseModel):
    """Inbound payload for recording a tournament-attempt result.

    ``status`` is a client-facing Literal that excludes ``"started"`` (the
    only pre-submit state) — a client can only ever submit a terminal result.
    ``honesty`` is deliberately absent: it is server-only and stays "pending"
    for every attempt created by this brick.
    """

    model_config = ConfigDict(extra="forbid")

    time_ms: int = Field(gt=0)
    status: Literal["valid", "dnf"] = "valid"


class TournamentAttemptRead(BaseModel):
    """Public representation of an attempt, merged with its tournament's
    week/scramble fields. The scramble is populated ONLY on the authed
    start/submit responses this schema backs (П8 — never a public route).
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    tournament_id: UUID
    status: str
    honesty: str
    time_ms: int | None
    started_at: datetime
    submitted_at: datetime | None
    iso_year: int
    iso_week: int
    week_label: str
    event: str
    scramble: str
    # Badges newly granted by THIS submit (not the caller's full history) —
    # best-effort, see app.services.badges / app.routers.tournament. Empty on
    # the `start` response (no evaluation happens there).
    new_badges: list[BadgeRead] = []


class StandingEntry(BaseModel):
    """One row on the de-ranked participation board.

    Deliberately has NO rank/position field (true ranking waits for the
    honesty-verification brick) and NO email/nickname — ``display_name`` is
    always ``public_handle`` or the literal "Аноним"
    (``app.services.tournament.display_name_for``).
    """

    model_config = ConfigDict(from_attributes=True)

    display_name: str
    time_ms: int
    is_self: bool


class TournamentStandingsRead(BaseModel):
    """Current-ISO-week participation board: `status=="valid"` completers
    ordered by `submitted_at ASC, id ASC`. Read-only (creates nothing, no
    scramble). Absent/empty tournament -> empty entries + zero counts, not 404.
    """

    model_config = ConfigDict(from_attributes=True)

    iso_year: int
    iso_week: int
    week_label: str
    event: str
    entries: list[StandingEntry]
    your_entry: StandingEntry | None
    valid_count: int
    dnf_count: int


class TournamentCurrentRead(BaseModel):
    """Read-only state of the caller's current-week tournament, deliberately
    WITHOUT a ``scramble`` field — this schema backs the ``GET
    /tournament/current`` endpoint, the one route that MUST NEVER reveal the
    weekly scramble (П8). It never triggers a create: ``attempt_status`` and
    the attempt-derived fields are ``None`` when no tournament/attempt exists
    yet for this user this week.
    """

    model_config = ConfigDict(from_attributes=True)

    iso_year: int
    iso_week: int
    week_label: str
    event: str
    attempt_status: AttemptStatus | None
    time_ms: int | None
    started_at: datetime | None
    submitted_at: datetime | None
    deadline_at: datetime | None

File: frontend/src/api/tournament.ts
// Weekly-tournament endpoints. Mirrors backend/app/schemas/tournament.py.
// Authed only — no public/anon route exists here, and none may: the shared
// weekly scramble is revealed ONLY by the two POST responses below, never by
// GET /tournament/current (П8 — see useTournamentAttempt.ts).

import { request } from "./client";
import type { BadgeRead } from "./badges";

export type TournamentAttemptStatus = "started" | "valid" | "dnf";

// GET /tournament/current — read-only snapshot, deliberately has NO `scramble`
// field on the wire. `attempt_status` is null when the caller has no attempt
// this week (the GET never creates a tournament/attempt row).
export interface TournamentCurrentRead {
  iso_year: number;
  iso_week: number;
  week_label: string;
  event: string;
  attempt_status: TournamentAttemptStatus | null;
  time_ms: number | null;
  started_at: string | null;
  submitted_at: string | null;
  deadline_at: string | null;
}

// POST .../attempt/start and .../attempt/submit both return the full attempt,
// INCLUDING `scramble` — this is the only shape that ever carries it.
export interface TournamentAttemptRead {
  id: string;
  tournament_id: string;
  status: string;
  honesty: string;
  time_ms: number | null;
  started_at: string;
  submitted_at: string | null;
  iso_year: number;
  iso_week: number;
  week_label: string;
  event: string;
  scramble: string;
  /** Badges newly granted by THIS submission (best-effort award engine); empty/absent otherwise. */
  new_badges?: BadgeRead[];
}

export interface TournamentAttemptSubmit {
  time_ms: number;
  status: "valid" | "dnf";
}

// GET /tournament/current/standings — de-ranked participation board (no rank/
// position field on the wire, by design — see TournamentStandings.tsx). Never
// carries email, only the opt-in `handle` (surfaced here as `display_name`,
// already "Аноним"-substituted server-side).
export interface StandingEntry {
  display_name: string;
  time_ms: number;
  is_self: boolean;
}

export interface TournamentStandingsRead {
  iso_year: number;
  iso_week: number;
  week_label: string;
  event: string;
  entries: StandingEntry[];
  your_entry: StandingEntry | null;
  valid_count: number;
  dnf_count: number;
}

export function getCurrent(signal?: AbortSignal): Promise<TournamentCurrentRead> {
  return request<TournamentCurrentRead>("/tournament/current", { signal });
}

export function getStandings(
  limit?: number,
  signal?: AbortSignal,
): Promise<TournamentStandingsRead> {
  const query = limit != null ? `?limit=${encodeURIComponent(String(limit))}` : "";
  return request<TournamentStandingsRead>(`/tournament/current/standings${query}`, { signal });
}

// Idempotent: a second call in the same week reloads the existing attempt (and
// the same shared scramble) instead of re-rolling.
export function startAttempt(signal?: AbortSignal): Promise<TournamentAttemptRead> {
  return request<TournamentAttemptRead>("/tournament/current/attempt/start", {
    method: "POST",
    signal,
  });
}

export function submitAttempt(
  body: TournamentAttemptSubmit,
  signal?: AbortSignal,
): Promise<TournamentAttemptRead> {
  return request<TournamentAttemptRead>("/tournament/current/attempt/submit", {
    json: body,
    signal,
  });
}
